clustered/client/src/main/java/org/ehcache/clustered/client/internal/service/ConnectionState.java

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.clustered.client.internal.service;
+
+import org.ehcache.CachePersistenceException;
+import org.ehcache.clustered.client.config.ClusteringServiceConfiguration;
+import org.ehcache.clustered.client.config.Timeouts;
+import org.ehcache.clustered.client.internal.ClusterTierManagerClientEntity;
+import org.ehcache.clustered.client.internal.ClusterTierManagerClientEntityFactory;
+import org.ehcache.clustered.client.internal.ClusterTierManagerCreationException;
+import org.ehcache.clustered.client.internal.ClusterTierManagerValidationException;
+import org.ehcache.clustered.client.internal.store.ClusterTierClientEntity;
+import org.ehcache.clustered.client.service.EntityBusyException;
+import org.ehcache.clustered.common.internal.ServerStoreConfiguration;
+import org.ehcache.clustered.common.internal.exceptions.DestroyInProgressException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terracotta.connection.Connection;
+import org.terracotta.connection.ConnectionException;
+import org.terracotta.connection.ConnectionPropertyNames;
+import org.terracotta.exception.ConnectionClosedException;
+import org.terracotta.exception.EntityAlreadyExistsException;
+import org.terracotta.exception.EntityNotFoundException;
+import org.terracotta.lease.connection.LeasedConnectionFactory;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Properties;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeoutException;
+
+class ConnectionState {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ConnectionState.class);
+
+  private static final String CONNECTION_PREFIX = "Ehcache:";
+
+  private volatile Connection clusterConnection = null;
```

Comment:
It seem weird that these things are volatile and the `ConnectionState` reference inside `DefaultClusteringService` was final. I had expected things to be the other way round.

Reply:
I know its weird, need to put more though there